### PR TITLE
feat: extract HTTP backoff utilities to lib/import/http-utils.ts

### LINKED ISSUE #158: refactor: Decompose fetchWithBackoff in open5eAdapter
## Problem
`lib/import/open5eAdapter.ts` `fetchWithBackoff()` (lines 72-114) has cognitive complexity ~6 due to handling URL validation, retry logic, backoff calculation, and 429 response handling all in one function.

## Solution
Decompose into smaller pure functions:
- Extract backoff calculation: `calculateBackoffMs(attempt: number, retryAfter?: string | null): number`
- Extract 429 handling: `handleRateLimitResponse(response: Response, attempt: number, retries: number): Promise<void>`
- Keep `fetchWithBackoff` as orchestrator

## Acceptance Criteria
- [ ] `calculateBackoffMs()` extracted
- [ ] `fetchWithBackoff` simplified to orchestration only
- [ ] Tests still pass
- [ ] File compiles without errors

### PR DESCRIPTION
**Pull request by dougis**
## Summary

Extracts `calculateBackoffMs` and `handleRateLimitResponse` from the inline backoff logic in `fetchWithBackoff` (`lib/import/open5eAdapter.ts`) into a new shared module `lib/import/http-utils.ts`.

closes #158

## What Changed

- **New file**: `lib/import/http-utils.ts` — exports `calculateBackoffMs`, `handleRateLimitResponse`; file-scoped `MAX_BACKOFF_MS = 10_000`
- **Modified**: `lib/import/open5eAdapter.ts` — `fetchWithBackoff` now delegates to the utilities; magic literal `10000` removed
- **New tests**: `tests/unit/import/http-utils.test.ts` — 13 unit tests covering all spec scenarios

## Behaviour

No observable behaviour change. Same retry count, same backoff math, same 10 000 ms cap. Existing adapter tests pass without modification.

## Test Coverage

- `calculateBackoffMs`: exponential growth (attempts 0–5), cap boundary, Retry-After header parsing (within cap, over cap, zero)
- `handleRateLimitResponse`: mid-sequence sleep + true, Retry-After header path, final-attempt no-sleep + false

### CODE REVIEW

File: lib/import/http-utils.ts [hunk 1]
```diff
@@ -0,0 +1,25 @@
+const MAX_BACKOFF_MS = 10_000;
+
+export function calculateBackoffMs(
+  attempt: number,
+  retryAfterHeader?: string | null
+): number {
+  if (retryAfterHeader != null) {
+    return Math.min(parseInt(retryAfterHeader, 10) * 1000, MAX_BACKOFF_MS);
+  }
```
[the remark this answers is not in the document]
Comment (dougis) [reply lib/import/http-utils.ts]: Fixed in the latest commit. Now uses a truthy check (guards empty string), validates `!isNaN(parsed)` before using the value, and clamps with `Math.max(0, ...)` to prevent negative backoff from a malformed negative header. Falls through to exponential backoff if the header is absent, empty, or non-numeric.

File: lib/import/http-utils.ts [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+  attempt: number,
+  retries: number
+): Promise<boolean> {
+  if (attempt === retries) {
+    return false;
+  }
+  const retryAfter = response.headers.get("Retry-After");
+  const backoffMs = calculateBackoffMs(attempt, retryAfter);
+  await new Promise((resolve) => setTimeout(resolve, backoffMs));
+  return true;
```
[the remark this answers is not in the document]
Comment (dougis) [reply lib/import/http-utils.ts]: Fixed — changed `attempt === retries` to `attempt >= retries`. Added a test case for `attempt > retries` (attempt=5, retries=3) to verify no sleep and false return.

File: openspec/changes/extract-http-backoff-utils/tasks.md [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+### Task 4 — Run existing open5eAdapter tests
+
+- [x] Run `npm test -- tests/unit/import/open5eAdapter` (or equivalent)
+- [x] All existing tests must pass without modification
+
+## Validation
+
+- [x] Run full unit test suite: `npm test`
+- [x] Run type check: `tsc --noEmit` (zero errors)
+- [x] Run build: `npm run build` (succeeds)
```
[the remark this answers is not in the document]
Comment (dougis) [reply openspec/changes/extract-http-backoff-utils/tasks.md]: Fixed — all `npm test` references in tasks.md updated to `npm run test:unit` (and test path filter flags corrected for jest `--testPathPattern`).

File: openspec/changes/extract-http-backoff-utils/tasks.md [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+
+## Execution
+
+### Task 1 — Create `lib/import/http-utils.ts`
+
+- [x] Create `lib/import/http-utils.ts` with:
+  - `const MAX_BACKOFF_MS = 10_000` (file-scoped, not exported)
+  - `export function calculateBackoffMs(attempt: number, retryAfterHeader?: string | null): number`
+    - If `retryAfterHeader` is provided: `Math.min(parseInt(retryAfterHeader, 10) * 1000, MAX_BACKOFF_MS)`
+    - Otherwise: `Math.min(1000 * Math.pow(2, attempt), MAX_BACKOFF_MS)`
```
[the remark this answers is not in the document]
Comment (dougis) [reply openspec/changes/extract-http-backoff-utils/tasks.md:16]: Fixed — added `.openspec.yaml` with `schema: sdd-with-feedback-loop` and `created: 2026-05-21`, matching the format in other archived changes.

File: tests/unit/import/http-utils.test.ts [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+      expect(calculateBackoffMs(0, "5")).toBe(5000);
+    });
+
+    it('caps at MAX_BACKOFF_MS for "30"', () => {
+      expect(calculateBackoffMs(0, "30")).toBe(10000);
+    });
+
+    it('returns 0 for "0"', () => {
+      expect(calculateBackoffMs(0, "0")).toBe(0);
+    });
```
[the remark this answers is not in the document]
Comment (dougis) [reply tests/unit/import/http-utils.test.ts:47]: Added three edge-case tests covering the fallback path: empty string `""` (falsy → exponential), non-numeric string `"not-a-number"` (truthy but NaN → exponential), and whitespace `"   "` (truthy but NaN → exponential). All pass.

File: lib/import/open5eAdapter.ts [hunk 6]
```diff
@@ -85,27 +87,17 @@ async function fetchWithBackoff(
...
 
       return response;
     } catch (error) {
       lastError = error as Error;
       if (attempt < retries) {
-        const backoffMs = Math.min(1000 * Math.pow(2, attempt), 10000);
-        await new Promise((resolve) => setTimeout(resolve, backoffMs));
+        await new Promise((resolve) =>
+          setTimeout(resolve, calculateBackoffMs(attempt))
+        );
```
[the remark this answers is not in the document]
Comment (dougis) [reply lib/import/open5eAdapter.ts:100]: Added `"retries after network error and returns data on success"` test in `open5eAdapter.test.ts` — mocks fetch to throw `Error("Network failure")` on the first call and return 200 on the second. Verifies `setTimeout` was called with 1000 ms (the catch-block backoff path).

File: lib/import/open5eAdapter.ts [hunk 7]
```diff
@@ -85,27 +87,17 @@ async function fetchWithBackoff(
...
-          backoffMs = Math.min(parseInt(retryAfter, 10) * 1000, 10000);
-        } else {
-          backoffMs = Math.min(1000 * Math.pow(2, attempt), 10000);
-        }
-
-        if (attempt < retries) {
-          await new Promise((resolve) => setTimeout(resolve, backoffMs));
-          continue;
-        }
+        if (await handleRateLimitResponse(response, attempt, retries)) continue;
```
[the remark this answers is not in the document]
Comment (dougis) [reply lib/import/open5eAdapter.ts:90]: Added `"retries after 429 and returns data on success"` test in `open5eAdapter.test.ts` — mocks fetch to return 429 (no Retry-After) on the first call and 200 with data on the second. Verifies retry loop `continue` path works end-to-end and that `setTimeout` was called with 1000 ms.
